stdlib/public/runtime/GenericMetadataBuilder.cpp

```diff
@@ -0,0 +1,408 @@
+  InProcessReaderWriter readerWriter;
+  GenericMetadataBuilder builder{readerWriter};
+
+  auto result = builder.initializeGenericMetadata(
+      {asFullMetadata(metadata), -1u}, nullptr);
+  if (auto *error = result.getError()) {
+    fprintf(stderr, "swift_initializeGenericValueMetadata failed: %s",
+            error->cStr());
+    return false;
+  }
+  return true;
+}
+
+static BuilderErrorOr<size_t>
+genericValueDataExtraSize(const ValueTypeDescriptor *description,
+                          const GenericMetadataPattern *pattern) {
+  InProcessReaderWriter readerWriter;
+  GenericMetadataBuilder builder{readerWriter};
+  return builder.extraDataSize({description}, {pattern});
+}
+
+static void _swift_dumpMetadata(const Metadata *metadata) {
+  GenericMetadataBuilder<InProcessReaderWriter>::Dumper dumper(printf);
+  auto result = dumper.dumpMetadata({metadata});
+  if (auto *error = result.getError()) {
+    printf("error: %s", error->cStr());
+  }
+}
+
+template <typename T>
+static const T &unwrapVWTField(const T &field) {
+  return field;
+}
+
+static uint32_t unwrapVWTField(const ValueWitnessFlags &field) {
+  return field.getOpaqueValue();
+}
+
+static bool equalVWTs(const ValueWitnessTable *a, const ValueWitnessTable *b) {
+#define WANT_ONLY_REQUIRED_VALUE_WITNESSES
+#define FUNCTION_VALUE_WITNESS(LOWER_ID, UPPER_ID, RET, PARAMS)                \
+  if (a->LOWER_ID != b->LOWER_ID)                                              \
+    return false;
+#define VALUE_WITNESS(LOWER_ID, UPPER_ID)                                      \
+  if (unwrapVWTField(a->LOWER_ID) != unwrapVWTField(b->LOWER_ID))              \
+    return false;
+#include "swift/ABI/ValueWitness.def"
+
+  auto *enumA = dyn_cast<EnumValueWitnessTable>(a);
+  auto *enumB = dyn_cast<EnumValueWitnessTable>(b);
+  if (enumA == nullptr && enumB == nullptr) {
+    return true;
+  } else if (enumA != nullptr && enumB != nullptr) {
+#define WANT_ONLY_ENUM_VALUE_WITNESSES
+#define VALUE_WITNESS(LOWER_ID, UPPER_ID)                                      \
+  if (unwrapVWTField(enumA->LOWER_ID) != unwrapVWTField(enumB->LOWER_ID))      \
+    return false;
+#include "swift/ABI/ValueWitness.def"
+    return true;
+  } else {
+    // Only one of a and b is an enum table.
+    return false;
+  }
+}
+
+void swift::validateExternalGenericMetadataBuilder(
+    const Metadata *original, const TypeContextDescriptor *description,
+    const void *arguments) {
+  if (auto valueDescriptor = dyn_cast<ValueTypeDescriptor>(description)) {
+    if (valueDescriptor->isGeneric()) {
+      auto pattern = reinterpret_cast<GenericValueMetadataPattern *>(
+          valueDescriptor->getFullGenericContextHeader()
+              .DefaultInstantiationPattern.get());
+      auto extraDataSize = genericValueDataExtraSize(valueDescriptor, pattern);
+      if (auto *error = extraDataSize.getError()) {
+        printf("Failed to get extra data size: %s\n", error->cStr());
```

Comment:
Do these want to go to `stdout`, or `stderr`, or some logging thing? And should they have some prefix to indicate where they're coming from?

Reply:
I've redone this stuff to be a little more sensible. They all go to `stderr` now, and they all go through a `validationLog` function which prints a prefix on everything, with the exception of the metadata dumper which expects to be able to print partial lines and so doesn't like prefixes much.

I also decided that validation and the various logs aren't quite the same thing, so I changed it to make `SWIFT_DEBUG_VALIDATE_EXTERNAL_GENERIC_METADATA_BUILDER` an integer, where 1 does the validation silently unless there's an error, and 2 does full logging along with it.